### PR TITLE
reduce computations for insertions + path sampling factor

### PR DESCRIPTION
**Pull request by yammineramy**
We attempt here to reduce the time it takes thunder to insert a path into the database. This code works perfectly and has shown good results with dat files generated using it being tested on robot and running in customer sites.
Dat file generation takes now around 30 mins or less, as opposed to multiple hours (5 or 6).

### CODE REVIEW

File: src/ompl/tools/thunder/SPARSdb.h [hunk 1]
```diff
@@ -433,11 +433,21 @@ namespace ompl
...
 
+            void setPathSamplingFactor(const std::optional<int> pathSamplingFactor)
+            {
+                pathSamplingFactor_ = pathSamplingFactor;
+            }
+
+            void setPathSamplingFactor(const int pathSamplingFactor)
+            {
+                pathSamplingFactor_ = pathSamplingFactor;
+            }
```
Comment (jackievideira) [inline src/ompl/tools/thunder/SPARSdb.h:449]: do you need both of these methods? i think you could just keep this one, is there any time you'd call this with out an int argument?

File: src/ompl/tools/thunder/src/SPARSdb.cpp [hunk 2]
```diff
@@ -658,16 +658,20 @@ bool ompl::geometric::SPARSdb::addPathToRoadmap(const base::PlannerTerminationCo
...
                "%i",
                numGuards, solutionPath.getStateCount());
 
-    unsigned int n = 0;
-    const int n1 = solutionPath.getStateCount() - 1;
-    for (int i = 0; i < n1; ++i)
-        n += si_->getStateSpace()->validSegmentCount(solutionPath.getState(i), solutionPath.getState(i + 1));
-
-    solutionPath.interpolate(n);
+    if (pathSamplingFactor_.has_value()) {
```
Comment (jackievideira) [inline src/ompl/tools/thunder/src/SPARSdb.cpp:665]: you wrote a getter for this but didn't use it :sweat_smile:

File: src/ompl/tools/thunder/SPARSdb.h [hunk 3]
```diff
@@ -433,11 +433,21 @@ namespace ompl
...
 
-            void setUseCostInRoadmap_(const bool useCostInRoadmap)
+            void setUseCostInRoadmap(const bool useCostInRoadmap)
             {
                 useCostInRoadmap_ = useCostInRoadmap;
             }
 
+            void setPathSamplingFactor(const std::optional<int> pathSamplingFactor)
+            {
+                pathSamplingFactor_ = pathSamplingFactor;
```
[the remark this answers is not in the document]
Comment (jackievideira) [reply src/ompl/tools/thunder/SPARSdb.h:443]: nevermind i see `pathSamplingFactor_` is an optional

File: src/ompl/tools/thunder/SPARSdb.h (hunk 1 above)
Comment (yammineramy) [reply src/ompl/tools/thunder/SPARSdb.h:449]: good point... I think this one is the only one that we will call

**Review by jackievideira — APPROVED** [review]
lgtm style + plat

File: src/ompl/tools/thunder/SPARSdb.h [hunk 4]
```diff
@@ -846,6 +856,9 @@ namespace ompl
 
             double heuristicScaling_ {1.0};
 
+            /** \brief optional variable. If set, will sample each segment in the path with this many nodes and attempt to insert */
```
Comment (duanyutong) [inline src/ompl/tools/thunder/SPARSdb.h:859]: extra \?

File: src/ompl/tools/thunder/src/SPARSdb.cpp [hunk 5]
```diff
@@ -658,16 +658,20 @@ bool ompl::geometric::SPARSdb::addPathToRoadmap(const base::PlannerTerminationCo
...
+        solutionPath.interpolate(pathSamplingFactor_.value() * solutionPath.getStateCount());
+    } else {
+        unsigned int n = 0;
+        const int n1 = solutionPath.getStateCount() - 1;
+        for (int i = 0; i < n1; ++i) {
+            n += si_->getStateSpace()->validSegmentCount(solutionPath.getState(i), solutionPath.getState(i + 1));
+        }
+        solutionPath.interpolate(n);
+    }
```
Comment (duanyutong) [inline src/ompl/tools/thunder/src/SPARSdb.cpp:674]: why does `pathSamplingFactor_` skip the else block?

File: src/ompl/tools/thunder/src/SPARSdb.cpp [hunk 6]
```diff
@@ -658,16 +658,20 @@ bool ompl::geometric::SPARSdb::addPathToRoadmap(const base::PlannerTerminationCo
...
-    for (int i = 0; i < n1; ++i)
-        n += si_->getStateSpace()->validSegmentCount(solutionPath.getState(i), solutionPath.getState(i + 1));
-
-    solutionPath.interpolate(n);
+    if (pathSamplingFactor_.has_value()) {
+        solutionPath.interpolate(pathSamplingFactor_.value() * solutionPath.getStateCount());
+    } else {
+        unsigned int n = 0;
+        const int n1 = solutionPath.getStateCount() - 1;
+        for (int i = 0; i < n1; ++i) {
```
Comment (duanyutong) [inline src/ompl/tools/thunder/src/SPARSdb.cpp:670]: uniform init

File: src/ompl/tools/thunder/src/SPARSdb.cpp [hunk 7]
```diff
@@ -962,6 +966,9 @@ bool ompl::geometric::SPARSdb::addStateToRoadmap(const base::PlannerTerminationC
     if (denseRoadmap_) {
         //this added call to findGraphNeighbors is very inexpensive when granularity is small enough (which it is meant to be).
         findGraphNeighbors(qNew, gnbhd, vnbhd, granularity_);
+        if (getNumConnectedComponents() == 1 && vnbhd.size()) {
+            return false;
```
Comment (duanyutong) [inline src/ompl/tools/thunder/src/SPARSdb.cpp:970]: add comment on what this does

File: src/ompl/tools/thunder/SPARSdb.h (hunk 4 above)
Comment (yammineramy) [reply src/ompl/tools/thunder/SPARSdb.h:859]: This is usually how they denote brief in ompl I think

**Pull request merged**